tmpl/blog/introducing-solo5.md

```diff
@@ -0,0 +1,125 @@
+
+### Solo5: Building a Unikernel Base From Scratch
+
+At a high level, there are roughly 3 parts to building a unikernel
+base that runs on KVM/QEMU and supports Mirage:
+
+* **Typical kernel hardware initialization.** The kernel must know how
+  to loading things into memory at the desired locations, and prepare
+  the processor to operate in the correct mode (e.g., 64-bit).  Unlike
+  typical kernels, most setup is one-time and simplified.  The kernel
+  must set up a memory map, stack, interrupt vectors, and provide
+  primitives for basic memory allocation.  At its simplest, a
+  unikernel base kernel does not need to worry about user address
+  spaces, threads, or many other things typical kernels need.
+
+* **Interact with `virtio` devices.** `virtio` is a paravirtualized
+  device standard supported by some hypervisors, including KVM/QEMU
+  and Virtualbox.  As far as devices go, `virtio` devices are simple:
+  I was able to write (very simple/unoptimized) `virtio` drivers for
+  Solo5 drivers from scratch in C.  At some point it may be
+  interesting to write them in OCaml like the Xen device drivers in
+  Mirage, but for someone who doesn't know OCaml (like me) a simple C
+  implementation seemed like a good first step.  I should note that
+  even though the drivers themselves are written in C, to connect with
+  Mirage Solo5 does include some OCaml code to call out to the
+  drivers.
+
+* **Appropriately link Mirage binaries/build system.** A piece of
+  software called mirage-platform performs the binding between Mini-OS
+  and the rest of the Mirage stack.  Building a new unikernel base
+  means that this "cut point" will have lots of undefined dependencies
+  which can either be implemented in the new unikernel base, stubbed
+  out, or reused.  Other "cut points" involve device drivers: the
+  console, network and block devices.  Finally, the `mirage` tool
+  needs to output appropriate Makefiles for the new target and an
+  overall Makefile needs to put everything together.
+
+Each one of these steps carries complexity and gotchas and I have
+certainly made many mistakes when performing all of them.  The
+hardware initialization process is needlessly complex, and the overall
+Makefile reflects my ignorance of OCaml and its building and packaging
+systems.  It's a work in progress!
+
+### Next Steps and Getting Involved
+
+In addition to the aforementioned clean up, I'm currently exploring
+the boot time in this environment.  So far I've found that generating
+a bootable iso with GRUB as a bootloader and relying on QEMU to
+emulate BIOS calls to load the kernel is... inefficient.
+
+If you find the lowest layer of the unikernel interesting, please
+don't hesitate to contact me or get involved.  I've packaged the build
+and test environment for Solo5 into a Docker container to reduce the
+dependency burden in playing around with it.  Check out [the
+repo](https://github.com/djwillia/solo5/tree/mirage) for the full
+instructions!
```

Comment:
Perhaps also mention that you'll be speaking about Solo5 at the upcoming SCALE/CIF event (I'm not really sure what the 'official' name is)?  Can link off to http://wiki.xenproject.org/wiki/2016_Unikernels_and_More:_Cloud_Innovators_Forum_Schedule